staging/src/k8s.io/apiserver/pkg/endpoints/discovery/root.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/emicklei/go-restful"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apiserver/pkg/endpoints/handlers/negotiation"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+)
+
+type DiscoveryGroupManager interface {
+	AddAPIGroup(apiGroup metav1.APIGroup)
+	RemoveAPIGroup(groupName string)
+
+	WebService() *restful.WebService
+}
+
+// rootAPIsDiscoveryHandler creates a webservice serving api group discovery.
+// Note: during the server runtime apiGroups might change.
```

Comment:
> Is this comment appropriate here? If so, what is it saying exactly?

This list of API groups is not fixed and (at least currently) this is accomplished by mutating the handler itself.  You cannot safely cache this information.